File: engine/src/flutter/lib/web_ui/lib/src/engine/skwasm/skwasm_impl/filters.dart
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'dart:ffi';
import 'dart:typed_data';

import 'package:ui/src/engine.dart';
import 'package:ui/src/engine/skwasm/skwasm_impl.dart';
import 'package:ui/ui.dart' as ui;

abstract class SkwasmImageFilter implements BackendImageFilter {
  /// Returns the native WebAssembly handle for this image filter.
  ///
  /// Returns a null pointer (nullptr) if the filter is invalid (e.g. non-invertible matrix).
  ImageFilterHandle get nativeFilter => _handle;
  ImageFilterHandle _handle = nullptr;

  @override
  void dispose() {
    if (_handle != nullptr) {
      imageFilterDispose(_handle);
      _handle = nullptr;
    }
  }

  @override
  ui.Rect filterBounds(ui.Rect inputBounds) {
    if (_handle == nullptr) {
      return inputBounds;
    }
    return withStackScope((StackScope scope) {
      final RawIRect rawRect = scope.convertIRectToNative(inputBounds);
      // The skwasm C++ backend allocates a bounding box and writes to the memory address
      // of `rawRect` safely in-place without returning a nullable pointer.
      imageFilterGetFilterBounds(_handle, rawRect);
      return scope.convertIRectFromNative(rawRect);
    });
  }
}

class SkwasmColorFilterImageFilter extends SkwasmImageFilter {
  SkwasmColorFilterImageFilter(BackendColorFilter colorFilter) {
    _handle = imageFilterCreateFromColorFilter((colorFilter as SkwasmColorFilter).handle);
  }
}

class SkwasmBlurImageFilter extends SkwasmImageFilter {
  SkwasmBlurImageFilter({
    required double sigmaX,
    required double sigmaY,
    required ui.TileMode tileMode,
  }) {
    _handle = imageFilterCreateBlur(sigmaX, sigmaY, tileMode.index);
  }
}

class SkwasmMatrixImageFilter extends SkwasmImageFilter {
  SkwasmMatrixImageFilter({required Float64List matrix, required ui.FilterQuality filterQuality}) {
    _handle = withStackScope((StackScope scope) {
      final Pointer<Float> matrixPtr = scope.convertMatrix4toSkMatrix(matrix);
      return imageFilterCreateMatrix(matrixPtr, filterQuality.index);
    });
  }
}

class SkwasmDilateImageFilter extends SkwasmImageFilter {
  SkwasmDilateImageFilter({required double radiusX, required double radiusY}) {
    _handle = imageFilterCreateDilate(radiusX, radiusY);
  }
}

class SkwasmErodeImageFilter extends SkwasmImageFilter {
  SkwasmErodeImageFilter({required double radiusX, required double radiusY}) {
    _handle = imageFilterCreateErode(radiusX, radiusY);
  }
}

class SkwasmComposeImageFilter extends SkwasmImageFilter {
  SkwasmComposeImageFilter({required BackendImageFilter outer, required BackendImageFilter inner}) {
    final ImageFilterHandle outerHandle = (outer as SkwasmImageFilter).nativeFilter;
    final ImageFilterHandle innerHandle = (inner as SkwasmImageFilter).nativeFilter;
    _handle = imageFilterCompose(outerHandle, innerHandle);
  }
}

/// Skwasm implementation of [ui.ColorFilter].
///
/// This class acts as a thin wrapper around a [ColorFilterHandle].
/// The memory lifecycle of the underlying native object is managed by the
/// [FinalizationRegistry] associated with the shared `EngineColorFilter`.
class SkwasmColorFilter implements BackendColorFilter {
  SkwasmColorFilter(EngineColorFilter colorFilter) : handle = _initHandle(colorFilter);

  final ColorFilterHandle handle;

  static ColorFilterHandle _initHandle(EngineColorFilter colorFilter) {
    switch (colorFilter.type) {
      case ColorFilterType.mode:
        final ui.Color? color = colorFilter.color;
        final ui.BlendMode? blendMode = colorFilter.blendMode;
        if (color == null || blendMode == null) {
          throw StateError('ColorFilter.mode must have a color and blend mode.');
        }
        return colorFilterCreateMode(color.value, blendMode.index);
      case ColorFilterType.linearToSrgbGamma:
        return colorFilterCreateLinearToSRGBGamma();
      case ColorFilterType.srgbToLinearGamma:
        return colorFilterCreateSRGBToLinearGamma();
      case ColorFilterType.matrix:
        final List<double>? matrix = colorFilter.matrix;
        if (matrix == null) {
          throw StateError('ColorFilter.matrix must have a matrix.');
        }
        return withStackScope((scope) {
          final Pointer<Float> rawMatrix = scope.convertDoublesToNative(matrix);
          const translationIndices = <int>[4, 9, 14, 19];
          for (final i in translationIndices) {
            // Flutter documentation says the translation column of the color matrix
            // is specified in unnormalized 0..255 space. Skia expects the
            // translation values to be normalized to 0..1 space.
            rawMatrix[i] /= 255.0;
          }
          return colorFilterCreateMatrix(rawMatrix);
        });
    }
  }

  bool _isDisposed = false;

  @override
  void dispose() {
    if (_isDisposed) {
      return;
    }
    _isDisposed = true;
    colorFilterDispose(handle);
  }
}

class SkwasmMaskFilter implements BackendMaskFilter {
  SkwasmMaskFilter(EngineMaskFilter maskFilter)
    : handle = maskFilterCreateBlur(maskFilter.webOnlyBlurStyle.index, maskFilter.webOnlySigma);

  final MaskFilterHandle handle;

  bool _isDisposed = false;

  @override
  void dispose() {
    if (_isDisposed) {
      return;
    }
    _isDisposed = true;
    maskFilterDispose(handle);
  }
}

File: engine/src/flutter/skwasm/filters.cc
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "flutter/display_list/effects/dl_color_filter.h"
#include "flutter/display_list/effects/dl_image_filter.h"
#include "flutter/display_list/effects/dl_mask_filter.h"
#include "flutter/skwasm/export.h"
#include "flutter/skwasm/helpers.h"
#include "flutter/skwasm/live_objects.h"

SKWASM_EXPORT Skwasm::sp_wrapper<flutter::DlImageFilter>*
imageFilter_createBlur(flutter::DlScalar sigma_x,
                       flutter::DlScalar sigma_y,
                       flutter::DlTileMode tile_mode) {
  Skwasm::live_image_filter_count++;
  return new Skwasm::sp_wrapper<flutter::DlImageFilter>(
      flutter::DlImageFilter::MakeBlur(sigma_x, sigma_y, tile_mode));
}

SKWASM_EXPORT Skwasm::sp_wrapper<flutter::DlImageFilter>*
imageFilter_createDilate(flutter::DlScalar radius_x,
                         flutter::DlScalar radius_y) {
  Skwasm::live_image_filter_count++;
  return new Skwasm::sp_wrapper<flutter::DlImageFilter>(
      flutter::DlImageFilter::MakeDilate(radius_x, radius_y));
}

SKWASM_EXPORT Skwasm::sp_wrapper<flutter::DlImageFilter>*
imageFilter_createErode(flutter::DlScalar radius_x,
                        flutter::DlScalar radius_y) {
  Skwasm::live_image_filter_count++;
  return new Skwasm::sp_wrapper<flutter::DlImageFilter>(
      flutter::DlImageFilter::MakeErode(radius_x, radius_y));
}

SKWASM_EXPORT Skwasm::sp_wrapper<flutter::DlImageFilter>*
imageFilter_createMatrix(flutter::DlScalar* matrix_33,
                         Skwasm::FilterQuality quality) {
  Skwasm::live_image_filter_count++;
  return new Skwasm::sp_wrapper<flutter::DlImageFilter>(
      flutter::DlImageFilter::MakeMatrix(
          Skwasm::CreateDlMatrixFrom3x3(matrix_33),
          Skwasm::SamplingOptionsForQuality(quality)));
}

SKWASM_EXPORT Skwasm::sp_wrapper<flutter::DlImageFilter>*
imageFilter_createFromColorFilter(
    Skwasm::sp_wrapper<flutter::DlColorFilter>* filter) {
  Skwasm::live_image_filter_count++;
  return new Skwasm::sp_wrapper<flutter::DlImageFilter>(
      flutter::DlImageFilter::MakeColorFilter(filter->Shared()));
}

SKWASM_EXPORT Skwasm::sp_wrapper<flutter::DlImageFilter>* imageFilter_compose(
    Skwasm::sp_wrapper<flutter::DlImageFilter>* outer,
    Skwasm::sp_wrapper<flutter::DlImageFilter>* inner) {
  Skwasm::live_image_filter_count++;
  return new Skwasm::sp_wrapper<flutter::DlImageFilter>(
      flutter::DlImageFilter::MakeCompose(outer->Shared(), inner->Shared()));
}

SKWASM_EXPORT void imageFilter_dispose(
    Skwasm::sp_wrapper<flutter::DlImageFilter>* filter) {
  Skwasm::live_image_filter_count--;
  delete filter;
}

SKWASM_EXPORT void imageFilter_getFilterBounds(
    Skwasm::sp_wrapper<flutter::DlImageFilter>* filter,
    flutter::DlIRect* in_out_bounds) {
  auto dl_filter = filter->Shared();
  if (dl_filter == nullptr) {
    // If there is no filter, the output bounds are the same as the input
    // bounds.
    return;
  }
  flutter::DlIRect in_rect = *in_out_bounds;
  dl_filter->map_device_bounds(in_rect, flutter::DlMatrix(), *in_out_bounds);
}

SKWASM_EXPORT Skwasm::sp_wrapper<const flutter::DlColorFilter>*
colorFilter_createMode(uint32_t color, flutter::DlBlendMode mode) {
  Skwasm::live_color_filter_count++;
  return new Skwasm::sp_wrapper<const flutter::DlColorFilter>(
      flutter::DlColorFilter::MakeBlend(flutter::DlColor(color), mode));
}

SKWASM_EXPORT Skwasm::sp_wrapper<const flutter::DlColorFilter>*
colorFilter_createMatrix(float* matrix_data  // 20 values
) {
  Skwasm::live_color_filter_count++;
  return new Skwasm::sp_wrapper<const flutter::DlColorFilter>(
      flutter::DlColorFilter::MakeMatrix(matrix_data));
}

SKWASM_EXPORT Skwasm::sp_wrapper<const flutter::DlColorFilter>*
colorFilter_createSRGBToLinearGamma() {
  Skwasm::live_color_filter_count++;
  return new Skwasm::sp_wrapper<const flutter::DlColorFilter>(
      flutter::DlColorFilter::MakeSrgbToLinearGamma());
}

SKWASM_EXPORT Skwasm::sp_wrapper<const flutter::DlColorFilter>*
colorFilter_createLinearToSRGBGamma() {
  Skwasm::live_color_filter_count++;
  return new Skwasm::sp_wrapper<const flutter::DlColorFilter>(
      flutter::DlColorFilter::MakeLinearToSrgbGamma());
}

SKWASM_EXPORT void colorFilter_dispose(
    Skwasm::sp_wrapper<const flutter::DlColorFilter>* filter) {
  Skwasm::live_color_filter_count--;
  delete filter;
}

SKWASM_EXPORT Skwasm::sp_wrapper<flutter::DlMaskFilter>* maskFilter_createBlur(
    flutter::DlBlurStyle blur_style,
    flutter::DlScalar sigma) {
  Skwasm::live_mask_filter_count++;
  return new Skwasm::sp_wrapper<flutter::DlMaskFilter>(
      flutter::DlBlurMaskFilter::Make(blur_style, sigma));
}

SKWASM_EXPORT void maskFilter_dispose(
    Skwasm::sp_wrapper<flutter::DlMaskFilter>* filter) {
  Skwasm::live_mask_filter_count--;
  delete filter;
}
